Test failure (22861) TestTransferRaftLeadership

Found with: https://circleci.com/gh/cockroachdb/cockroach/22861

```
+ echo 'FAIL: excerpt.txt is not empty (/tmp/circle-artifacts.ZLmtT6t/excerpt.txt)'
FAIL: excerpt.txt is not empty (/tmp/circle-artifacts.ZLmtT6t/excerpt.txt)
+ echo

+ head -n 100 /tmp/circle-artifacts.ZLmtT6t/excerpt.txt
I160916 03:51:13.586814 15234 storage/engine/rocksdb.go:404  closing in-memory rocksdb instance
I160916 03:51:13.586967 15234 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
I160916 03:51:13.586980 15234 storage/engine/rocksdb.go:404  closing in-memory rocksdb instance
I160916 03:51:13.587110 15234 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
I160916 03:51:13.587123 15234 storage/engine/rocksdb.go:404  closing in-memory rocksdb instance
--- FAIL: TestTransferRaftLeadership (0.07s)
	client_raft_test.go:2425: range 2 was not found
```

Reliably fails on a 24-CPU GCE worker with `-p 100`.

```
I160921 17:20:18.714212 1 rand.go:76  Random seed: 6025495756466001003
I160921 17:20:18.776372 34 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I160921 17:20:18.785830 34 gossip/gossip.go:230  initial resolvers: []
W160921 17:20:18.785885 34 gossip/gossip.go:1041  no resolvers found; use --join to specify a connected node
I160921 17:20:18.786045 34 gossip/gossip.go:264  NodeID set to 1
I160921 17:20:18.952436 34 gossip/gossip.go:264  NodeID set to 1
I160921 17:20:18.953216 34 storage/store.go:1018  [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I160921 17:20:18.953333 34 gossip/gossip.go:278  NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:38574" > attrs:<>
I160921 17:20:18.955323 75 storage/replica_trigger.go:254  [s1,r1{/Min-/Max}] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 8h20m0s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=1970-01-01 00:00:00 +0000 UTC]
I160921 17:20:18.955899 34 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I160921 17:20:18.960917 34 gossip/gossip.go:230  initial resolvers: [%!s(*resolver.socketResolver=&{tcp 127.0.0.1:38574})]
W160921 17:20:18.960953 34 gossip/gossip.go:1043  no incoming or outgoing connections
I160921 17:20:18.961029 34 gossip/gossip.go:264  NodeID set to 2
I160921 17:20:18.962050 34 gossip/gossip.go:264  NodeID set to 2
I160921 17:20:18.962280 34 storage/store.go:1018  [n2,s2]: failed initial metrics computation: [n2,s2]: system config not yet available
I160921 17:20:18.989479 34 gossip/gossip.go:278  NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:33185" > attrs:<>
I160921 17:20:18.989739 34 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I160921 17:20:18.990708 34 gossip/gossip.go:230  initial resolvers: [%!s(*resolver.socketResolver=&{tcp 127.0.0.1:38574})]
W160921 17:20:18.990733 34 gossip/gossip.go:1043  no incoming or outgoing connections
I160921 17:20:18.990775 34 gossip/gossip.go:264  NodeID set to 3
I160921 17:20:18.991722 21 gossip/client.go:125  node 2: started gossip client to 127.0.0.1:38574
I160921 17:20:18.992972 34 gossip/gossip.go:264  NodeID set to 3
I160921 17:20:18.993724 34 storage/store.go:1018  [n3,s3]: failed initial metrics computation: [n3,s3]: system config not yet available
I160921 17:20:18.993768 34 gossip/gossip.go:278  NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:46413" > attrs:<>
I160921 17:20:18.994269 335 gossip/client.go:125  node 3: started gossip client to 127.0.0.1:38574
I160921 17:20:19.010724 372 storage/replica_command.go:2304  [s1,r1{/Min-/Max}] initiating a split of this range at key "a"
I160921 17:20:19.017787 380 storage/replica_raftstorage.go:490  generated snapshot for range 2 at index 11 in 200.801µs. encoded size=3549, 28 KV pairs, 1 log entries
I160921 17:20:19.018801 386 storage/replica_raftstorage.go:629  [n2,s2,r2{-}]: with replicaID [?], applying preemptive snapshot at index 11 (encoded size=3549, 28 KV pairs, 1 log entries)
I160921 17:20:19.020244 386 storage/replica_raftstorage.go:632  [n2,s2,r2{"a"-/Max}]: with replicaID [?], applied preemptive snapshot in 0.001s
E160921 17:20:19.020702 15 storage/queue.go:543  [replicate] (purgatory) on [n1,s1,r1{/Min-"a"}]: 0 of 0 stores with an attribute matching []; likely not enough nodes in cluster
E160921 17:20:19.020883 15 storage/queue.go:543  [replicate] (purgatory) on [n1,s1,r2{"a"-/Max}]: 0 of 0 stores with an attribute matching []; likely not enough nodes in cluster
I160921 17:20:19.027274 34 storage/replica_command.go:3216  change replicas: read existing descriptor range_id:2 start_key:"a" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2
W160921 17:20:19.048724 340 storage/stores.go:212  range not contained in one range: [/Meta2/Max,"a\x00"), but have [/Min,"a")
W160921 17:20:19.114368 343 storage/intent_resolver.go:309  [n1,s1,r1{/Min-"a"}]: failed to push during intent resolution: failed to push "change-replica" id=8c6629db key=/Local/Range/"a"/RangeDescriptor rw=true pri=0.02017748 iso=SERIALIZABLE stat=PENDING epo=0 ts=0.000000000,74 orig=0.000000000,74 max=0.000000000,74 wto=false rop=false
I160921 17:20:19.115753 344 storage/replica.go:1688  [s1,r2{"a"-/Max}] proposing ADD_REPLICA {NodeID:2 StoreID:2 ReplicaID:2} for range 2: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2}]
I160921 17:20:19.122169 27 storage/replica_raftstorage.go:490  generated snapshot for range 2 at index 14 in 267.686µs. encoded size=4577, 30 KV pairs, 4 log entries
I160921 17:20:19.123681 402 storage/replica_raftstorage.go:629  [n3,s3,r2{-}]: with replicaID [?], applying preemptive snapshot at index 14 (encoded size=4577, 30 KV pairs, 4 log entries)
I160921 17:20:19.125210 402 storage/replica_raftstorage.go:632  [n3,s3,r2{"a"-/Max}]: with replicaID [?], applied preemptive snapshot in 0.001s
I160921 17:20:19.126114 34 storage/replica_command.go:3216  change replicas: read existing descriptor range_id:2 start_key:"a" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:2 store_id:2 replica_id:2 > next_replica_id:3
I160921 17:20:19.252656 385 storage/replica.go:1688  [s1,r2{"a"-/Max}] proposing ADD_REPLICA {NodeID:3 StoreID:3 ReplicaID:3} for range 2: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2} {NodeID:3 StoreID:3 ReplicaID:3}]
I160921 17:20:19.253494 418 storage/raft_transport.go:395  raft transport stream to node 1 established
I160921 17:20:19.258599 423 util/stop/stopper.go:425  quiescing; tasks left:
1      storage/replica_range_lease.go:145
1      storage/intent_resolver.go:377
1      storage/client_test.go:498
I160921 17:20:19.259307 435 storage/raft_transport.go:395  raft transport stream to node 1 established
I160921 17:20:19.261557 423 util/stop/stopper.go:425  quiescing; tasks left:
1      storage/intent_resolver.go:377
1      storage/client_test.go:498
I160921 17:20:19.262708 423 util/stop/stopper.go:425  quiescing; tasks left:
1      storage/intent_resolver.go:377
I160921 17:20:19.262826 422 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
I160921 17:20:19.263042 422 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
W160921 17:20:19.264590 406 storage/raft_transport.go:434  no handler found for store 1 in response range_id:2 from_replica:<node_id:2 store_id:2 replica_id:2 > to_replica:<node_id:1 store_id:1 replica_id:1 > union:<error:<message:"storage/raft_transport.go:250: unable to accept Raft message from {NodeID:1 StoreID:1 ReplicaID:1}: no handler registered for {NodeID:2 StoreID:2 ReplicaID:2}" transaction_restart:NONE origin_node:0 now:<wall_time:0 logical:0 > > >
W160921 17:20:19.266214 437 storage/store.go:2589  [s3] got error from range 2, replica {1 1 1}: storage/raft_transport.go:250: unable to accept Raft message from {NodeID:3 StoreID:3 ReplicaID:3}: no handler registered for {NodeID:1 StoreID:1 ReplicaID:1}
I160921 17:20:19.291588 422 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
W160921 17:20:19.307678 420 storage/raft_transport.go:434  no handler found for store 2 in response range_id:2 from_replica:<node_id:1 store_id:1 replica_id:1 > to_replica:<node_id:2 store_id:2 replica_id:2 > union:<error:<message:"storage/raft_transport.go:250: unable to accept Raft message from {NodeID:2 StoreID:2 ReplicaID:2}: no handler registered for {NodeID:1 StoreID:1 ReplicaID:1}" transaction_restart:NONE origin_node:0 now:<wall_time:0 logical:0 > > >
I160921 17:20:19.307903 422 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
W160921 17:20:19.307958 420 storage/raft_transport.go:434  no handler found for store 2 in response range_id:2 from_replica:<node_id:1 store_id:1 replica_id:1 > to_replica:<node_id:2 store_id:2 replica_id:2 > union:<error:<message:"storage/raft_transport.go:250: unable to accept Raft message from {NodeID:2 StoreID:2 ReplicaID:2}: no handler registered for {NodeID:1 StoreID:1 ReplicaID:1}" transaction_restart:NONE origin_node:0 now:<wall_time:0 logical:0 > > >
W160921 17:20:19.308036 420 storage/raft_transport.go:434  no handler found for store 2 in response range_id:2 from_replica:<node_id:1 store_id:1 replica_id:1 > to_replica:<node_id:2 store_id:2 replica_id:2 > union:<error:<message:"storage/raft_transport.go:250: unable to accept Raft message from {NodeID:2 StoreID:2 ReplicaID:2}: no handler registered for {NodeID:1 StoreID:1 ReplicaID:1}" transaction_restart:NONE origin_node:0 now:<wall_time:0 logical:0 > > >
W160921 17:20:19.308102 420 storage/raft_transport.go:434  no handler found for store 2 in response range_id:2 from_replica:<node_id:1 store_id:1 replica_id:1 > to_replica:<node_id:2 store_id:2 replica_id:2 > union:<error:<message:"storage/raft_transport.go:250: unable to accept Raft message from {NodeID:2 StoreID:2 ReplicaID:2}: no handler registered for {NodeID:1 StoreID:1 ReplicaID:1}" transaction_restart:NONE origin_node:0 now:<wall_time:0 logical:0 > > >
W160921 17:20:19.308194 420 storage/raft_transport.go:434  no handler found for store 2 in response range_id:2 from_replica:<node_id:1 store_id:1 replica_id:1 > to_replica:<node_id:2 store_id:2 replica_id:2 > union:<error:<message:"storage/raft_transport.go:250: unable to accept Raft message from {NodeID:2 StoreID:2 ReplicaID:2}: no handler registered for {NodeID:1 StoreID:1 ReplicaID:1}" transaction_restart:NONE origin_node:0 now:<wall_time:0 logical:0 > > >
I160921 17:20:19.308479 252 http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:46413->127.0.0.1:47660: use of closed network connection
I160921 17:20:19.308537 10 http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:38574->127.0.0.1:33604: use of closed network connection
I160921 17:20:19.308632 422 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
I160921 17:20:19.308976 422 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
I160921 17:20:19.309157 422 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
--- FAIL: TestTransferRaftLeadership (0.59s)
	client_raft_test.go:2425: range 2 was not found
FAIL


ERROR: exit status 1

16535 runs completed, 1 failures, over 1m57s
FAIL
Makefile:146: recipe for target 'stress' failed
make: *** [stress] Error 1
```

--- a/storage/client_raft_test.go
+++ b/storage/client_raft_test.go
@@ -2443,6 +2443,8 @@ func TestTransferRaftLeadership(t *testing.T) {
 	}
 	mtc.Start(t, numStores)
 	defer mtc.Stop()
+	store0 := mtc.Store(0)
+	store1 := mtc.Store(1)
 
 	key := roachpb.Key("a")
 
@@ -2454,14 +2456,14 @@ func TestTransferRaftLeadership(t *testing.T) {
 		}
 	}
 
-	rng := mtc.stores[0].LookupReplica(keys.MustAddr(key), nil)
+	rng := store0.LookupReplica(keys.MustAddr(key), nil)
 	if rng == nil {
 		t.Fatalf("no replica found for key '%s'", key)
 	}
 	mtc.replicateRange(rng.RangeID, 1, 2)
 
 	getArgs := getArgs([]byte("a"))
-	if _, pErr := client.SendWrappedWith(mtc.stores[0], context.Background(), roachpb.Header{RangeID: rng.RangeID}, &getArgs); pErr != nil {
+	if _, pErr := client.SendWrappedWith(store0, context.Background(), roachpb.Header{RangeID: rng.RangeID}, &getArgs); pErr != nil {
 		t.Fatalf("expect get nil, actual get %v ", pErr)
 	}
 
@@ -2473,20 +2475,22 @@ func TestTransferRaftLeadership(t *testing.T) {
 	// Force a read on Store 2 to request a new lease. Other moving parts in
 	// the system could have requested another lease as well, so we
 	// expire-request in a loop until we get our foot in the door.
-	var pErr *roachpb.Error
 	for {
 		mtc.expireLeases()
-		if _, pErr = client.SendWrappedWith(
-			mtc.stores[1],
+		if _, pErr := client.SendWrappedWith(
+			store1,
 			context.Background(),
 			roachpb.Header{RangeID: rng.RangeID},
 			&getArgs,
-		); testutils.IsPError(pErr, "not lease holder") {
-			continue
-		} else if pErr != nil {
-			t.Fatal(pErr)
+		); pErr == nil {
+			break
+		} else {
+			switch pErr.GetDetail().(type) {
+			case *roachpb.NotLeaseHolderError, *roachpb.RangeNotFoundError:
+			default:
+				t.Fatal(pErr)
+			}
 		}
-		break
 	}
 	// Wait for raft leadership transferring to be finished.
 	util.SucceedsSoon(t, func() error {

